File: leetcode/GRIND-75/Subsets.cpp
/*https://leetcode.com/problems/subsets/description/*/

class Solution {
public:
    void getSubsets(vector<int>&nums, vector<vector<int>> &res, int idx, vector<int> subset){
        if(idx > nums.size())
            return;
        res.push_back(subset);
        for(int i = idx; i<nums.size(); i++){
            vector<int> p = subset;
            p.push_back(nums[i]);
            getSubsets(nums, res, i+1, p);
        }
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>> res;
        vector<int> temp;
        res.push_back(temp);
        for(int i=0; i<n; i++){
            vector<int> subset;
            subset.push_back(nums[i]);
            getSubsets(nums, res, i+1, subset);

        }
        return res;
    }
};

File: leetcode/GRIND-75/Subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        backtrack(nums, 0, new ArrayList<>(), result);
        return result;
    }

    private void backtrack(int[] nums, int start, List<Integer>tempList, List<List<Integer>> result){
        result.add(new ArrayList<>(tempList));
        for(int i = start; i<nums.length; i++){
            tempList.add(nums[i]);
            backtrack(nums, i+1, tempList, result);
            tempList.remove(tempList.size()-1);
        }

    }
}